README.md

```diff
@@ -647,14 +647,382 @@ type VirtualMachine struct {
+Once we retrive the felt, we have the `encoded insruction`, we need to decode it to get the fields from its bits representation. Felt is not useful anymore so we will get its integer representation. 
+Now its time to decode the instruction and then run the `decoded instruction`.
+
+
+```go
+func (v *VirtualMachine) Step() error {
+	encoded_instruction, err := v.Segments.Memory.Get(v.RunContext.Pc)
+	if err != nil {
+		return fmt.Errorf("Failed to fetch instruction at %+v", v.RunContext.Pc)
+	}
+
+	encoded_instruction_felt, ok := encoded_instruction.GetFelt()
+	if !ok {
+		return errors.New("Wrong instruction encoding")
+	}
+
+	encoded_instruction_uint, err := encoded_instruction_felt.ToU64()
+	if err != nil {
+		return err
+	}
+
+	instruction, err := DecodeInstruction(encoded_instruction_uint)
+	if err != nil {
+		return err
+	}
+
+	return v.RunInstruction(&instruction)
+}
+```
+
+#### Decode instruction 
+
+```go
+//  Structure of the 63-bit that form the first word of each instruction.
+//  See Cairo whitepaper, page 32 - https://eprint.iacr.org/2021/1063.pdf.
+// ┌─────────────────────────────────────────────────────────────────────────┐
+// │                     off_dst (biased representation)                     │
+// ├─────────────────────────────────────────────────────────────────────────┤
+// │                     off_op0 (biased representation)                     │
+// ├─────────────────────────────────────────────────────────────────────────┤
+// │                     off_op1 (biased representation)                     │
+// ├─────┬─────┬───────┬───────┬───────────┬────────┬───────────────────┬────┤
+// │ dst │ op0 │  op1  │  res  │    pc     │   ap   │      opcode       │ 0  │
+// │ reg │ reg │  src  │ logic │  update   │ update │                   │    │
+// ├─────┼─────┼───┬───┼───┬───┼───┬───┬───┼───┬────┼────┬────┬────┬────┼────┤
+// │  0  │  1  │ 2 │ 3 │ 4 │ 5 │ 6 │ 7 │ 8 │ 9 │ 10 │ 11 │ 12 │ 13 │ 14 │ 15 │
+// └─────┴─────┴───┴───┴───┴───┴───┴───┴───┴───┴────┴────┴────┴────┴────┴────┘
+```
+
+As we can see in the picture above, all the information we need is present on the bits representation of the instruction. The first thing to do is create a structure that stores it. 
+
+```go 
+type Instruction struct {
+	Off0     int
+	Off1     int
+	Off2     int
+	DstReg   Register
+	Op0Reg   Register
+	Op1Addr  Op1Src
+	ResLogic ResLogic
+	PcUpdate PcUpdate
+	ApUpdate ApUpdate
+	FpUpdate FpUpdate
+	Opcode   Opcode
+}
+```
+
+And the good thing about this, is every bits combination for each field is known, so we can code all the possible flags to work with. These flags are represented below. 
+
+```go 
+// x-----------------------------x
+// x----- Instruction flags -----x
+// x-----------------------------x
+
```

Comment:
Can we add a short explanation of what each flag represents?

Reply:
Done!